Wangscape/TerrainImages.cpp

```diff
@@ -46,9 +30,12 @@ void TerrainImages::addTerrain(TerrainID tid, std::string image_filename, std::s
         it = source_images.insert({ filename, img }).first;
     }
     sf::Vector2i offset(offset_x, offset_y);
+    sf::Image tile;
+    tile.create(resolution.x, resolution.y);
+    tile.copy(it->second, 0,0, sf::IntRect(offset, sf::Vector2i(resolution)));
+    rearrangeTile(tile);
     sf::Texture tex;
-    if (!tex.loadFromImage((*it).second, sf::IntRect(offset, sf::Vector2i(resolution))))
+    if (!tex.loadFromImage(tile))
         throw std::runtime_error("Couldn't make texture from image");
```

Comment:
I've extracted the BFS component of `distances` and `isConnected` into a function, but it uses four different callbacks. Perhaps it would be better to make it a functor with 4 protected pure virtual functions.